Minor touchups.

AshkanAliabadi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44017 Add Vulkan op Add.
* #44016 Add Vulkan Tensor factory.
* #44015 Add Vulkan Tensor.
* #45148 Add fence.
* #45037 Add image sampler.
* #44496 Redo Vulkan command and descriptor pools.
* #44059 Add architectural support for multi-GPU.
* **#44317 Minor touchups.**

Differential Revision: [D23820828](https://our.internmc.facebook.com/intern/diff/D23820828)

Commit:
Minor touchups.

[ghstack-poisoned]

Commit:
Update on "Minor touchups."




[ghstack-poisoned]

Commit:
Update on "Minor touchups."




[ghstack-poisoned]

Commit:
Update on "Minor touchups."




[ghstack-poisoned]

Commit:
Update on "Minor touchups."




[ghstack-poisoned]

Commit:
Update on "Minor touchups."




[ghstack-poisoned]

Commit:
Update on "Minor touchups."




[ghstack-poisoned]

Commit:
Update on "Minor touchups."




[ghstack-poisoned]

Commit:
Update on "Minor touchups."




[ghstack-poisoned]

Commit:
Update on "Minor touchups."




[ghstack-poisoned]

Commit:
Update on "Minor touchups."




[ghstack-poisoned]

Commit:
Update on "Minor touchups."




[ghstack-poisoned]

IvanKobzarev commented on aten/src/ATen/native/vulkan/api/Allocator.h:
```diff
@@ -2,11 +2,19 @@
+    do {                              \
+      printf(format, ##__VA_ARGS__);  \
+      printf("\n");                   \
+    } while(false)
```
Is it some common pattern to do this in the do-while(false) cycle?
In which cases will it run more than once?

AshkanAliabadi replied:
Yes, it is a common pattern.  It is done so as to allow macros to be used safely in all place including places where only a single statement is allowed.  This wiki has a good explanation (https://wiki.sei.cmu.edu/confluence/display/c/PRE10-C.+Wrap+multistatement+macros+in+a+do-while+loop) and contains other implementations that might come to mind but are flawed.  By the way, the rest of the advice in there looks solid as well.

AshkanAliabadi replied:
Forgot to mention that it does not run more than once.  Execution runs through the body and terminates after the first run since the condition is false.

Commit:
Update on "Minor touchups."




[ghstack-poisoned]

IvanKobzarev commented on aten/src/ATen/native/vulkan/api/Resource.h:
```diff
@@ -25,12 +25,25 @@ struct C10_EXPORT Resource final {
+    // encapsualted in a scoped map-unmap region.  Allowing below overloads
+    // to be invoked on a temporary would open the door to the possibility
+    // of accessing the underlying memory out of the expected scope making
+    // for seemingly ineffective memory writes and hard to hunt down bugs.
```
I am getting that map() called on the temporary expression will be mapped-unmapped during the lifetime of that expression.

But I can not understand how that unmap will outlive that expression? Is it somewhen the chain of rvalues or something else?

AshkanAliabadi replied:
Right, great observation.  I think you have a point there and this particular application of this restriction on this class is a gray area, I'll explain why.

At a high level, and as a guideline that I try to follow, when returning the internals of an object, we want to make sure the lifetime of the returned handle (reference, pointer, etc.) will not exceed that of the object itself, so that was the high level motivation.

With that said, the ultimate usage pattern I wanted to prevent was this:

```
// Below line is a function declaration.
// Imagine we have a function that returns a `Memory` object.
Memory get_memory();

// Now we do something like this:
float * const pointer = get_memory().map<float>().get();
pointer[0] = 1.0f;

// This write above occurs outside of a mapped region.
// On top of that, if the internals of 'Memory' did not live in a pool (which it currently does)
// or if Memory would release its resources at destruction time (which it currently does not)
// then the problem was even more fundamental than access outside of a map-unmap region
// in that the pointer would now be pointing to invalid memory since 'Memory' is now destructed.
```

OK great.  Now your point made me realize that limiting `Handle::get()` to lvalue reference only would have been enough to prevent this behavior considering that the contents of `Memory` live in a pool so they outlive `Memory` anyway.  As an aside, limiting `Handle::get()` to lvalue references is something I do in this PR as well.

With that said, if we ever change `Memory` to release its contents eagerly the moment it goes out of scope, then we will need these overrides again, because `map()`'s return value provides access to the internal data.  In other words, assuming that `Memory` contents were released at the point its destructor was called, which is not the case with the current implementation by the way, then we would have:

```
// Below line is a function declaration.
// Imagine we have a function that returns a `Memory` object.
Memory get_memory();

const Memory::Data<float> data = get_memory().map();

// At this point Memory is destructed.  Data<float> holds a handle to invalid memory,
// so this usage below is problematic:

const float * pointer = data.get();
printf("%f\n", pointer[0]);
```

The reason I said this particular application of this restriction is a gray area here is because since Memory contents are managed in the pool, they outlive `Memory` so above usage is not an issue today.

Commit:
Update on "Minor touchups."




[ghstack-poisoned]

Commit:
Update on "Minor touchups."




[ghstack-poisoned]

Commit:
Update on "Minor touchups."




[ghstack-poisoned]

Commit:
Update on "Minor touchups."


Differential Revision: [D23820828](https://our.internmc.facebook.com/intern/diff/D23820828)

[ghstack-poisoned]

Commit:
Update on "Minor touchups."


Differential Revision: [D23820828](https://our.internmc.facebook.com/intern/diff/D23820828)

[ghstack-poisoned]

Commit:
Update on "Minor touchups."


Differential Revision: [D23820828](https://our.internmc.facebook.com/intern/diff/D23820828)

[ghstack-poisoned]

Commit:
Update on "Minor touchups."


Differential Revision: [D23820828](https://our.internmc.facebook.com/intern/diff/D23820828)

[ghstack-poisoned]
